src/trace.zig

```diff
@@ -284,63 +194,73 @@ pub const Tracer = struct {
+            .event_tracing = event_tracing,
+            .event_timing = event_timing,
+            .args = std.json.Formatter(Event){ .value = event, .options = .{} },
         }) catch unreachable;
 
         writer.writeAll(buffer_stream.getWritten()) catch |err| {
             std.debug.panic("Tracer.start: {}\n", .{err});
         };
     }
 
-    pub fn stop(tracer: *Tracer, event: Event, data: anytype) void {
-        comptime assert(@typeInfo(@TypeOf(data)) == .Struct);
+    pub fn stop(tracer: *Tracer, event: Event) void {
+        const us_log_threshold_ns = 5 * std.time.ns_per_ms;
+
+        const event_tracing = event.as(EventTracing);
+        const event_timing = event.as(EventTiming);
+        const stack = event_tracing.stack();
 
-        const stack = event.stack();
         const event_start_ns = tracer.events_started[stack].?;
         const event_end_ns = tracer.timer.read();
+        const duration_ns = event_end_ns - event_start_ns;
+        const duration_us = @divFloor(duration_ns, std.time.ns_per_us);
 
         assert(tracer.events_started[stack] != null);
         tracer.events_started[stack] = null;
 
-        log.debug("{}: {}: stop:{} (duration={}ms)", .{
+        // Double leading space to align with 'start: '.
+        log.debug("{}: {s}({}): stop:  {} (duration={}{s})", .{
             tracer.replica_index,
-            event,
-            struct_format(data, .dense),
-            @divFloor(event_end_ns - event_start_ns, std.time.ns_per_ms),
+            @tagName(event),
+            event_tracing,
+            event_timing,
+            if (duration_ns < us_log_threshold_ns)
+                duration_us
+            else
+                @divFloor(duration_ns, std.time.ns_per_ms),
+            if (duration_ns < us_log_threshold_ns) "us" else "ms",
         });
 
-        tracer.write_stop(stack, data);
+        tracer.metrics.record_timing(event_timing, duration_us);
+
+        tracer.write_stop(stack);
     }
 
-    pub fn reset(tracer: *Tracer, event_tag: Event.EventTag) void {
-        const stack_base = Event.event_stack_base.get(event_tag);
-        const cardinality = Event.event_stack_cardinality.get(event_tag);
+    pub fn reset(tracer: *Tracer, event_tag: Event.Tag) void {
+        const stack_base = EventTracing.stack_bases.get(event_tag);
+        const cardinality = EventTracing.stack_limits.get(event_tag);
         for (stack_base..stack_base + cardinality) |stack| {
             if (tracer.events_started[stack]) |_| {
                 log.debug("{}: {s}: reset", .{ tracer.replica_index, @tagName(event_tag) });
 
                 tracer.events_started[stack] = null;
-                tracer.write_stop(@intCast(stack), .{});
+                tracer.write_stop(@intCast(stack));
             }
         }
-    }
 
-    fn write_stop(tracer: *Tracer, stack: u32, data: anytype) void {
-        comptime assert(@typeInfo(@TypeOf(data)) == .Struct);
+        tracer.metrics.reset_all();
```

Comment:
:thinking: Any idea why this reset is here?

Reply:
Probably just an accident!

Maybe the bug is that `Tracer.rest` is called reset rather than cancel? It is actively doing things, it's not simply resetting the memory. 

I am not sure what's our naming scheme for reset vs cancel, but my decision tree is:

* if it is async, then definitely cancel. 
* if it's simple data structure `clear` than it's reset.
* if it is synchronous, but is non-trivial in some way (side effectual, or actively removes resources from unrelated links lists), than it also should be cancel. 

Thinking bigger, maybe _nothing_ should be called `reset` and instead we should use `clear` for data structures and `cancel` for everything else?